clarity/src/vm/functions/principals.rs

```diff
@@ -38,13 +69,259 @@ pub fn special_is_standard(
+    }
+
+    // Construct the principal.
+    let mut transfer_buffer = [0u8; 20];
+    transfer_buffer.copy_from_slice(&verified_hash_bytes);
+    let principal_data = StandardPrincipalData(version_byte, transfer_buffer);
+
+    let principal = if let Some(name) = name_opt {
+        // requested a contract principal.  Verify that the `name` is a valid ContractName.
+        // The type-checker will have verified that it's (string-ascii 40), but not long enough.
+        let name_bytes = match name {
+            Value::Sequence(SequenceData::String(CharType::ASCII(ref ascii_data))) => {
+                &ascii_data.data
+            }
+            _ => {
+                return Err(CheckErrors::TypeValueError(
+                    TypeSignature::contract_name_string_ascii(),
+                    name,
+                )
+                .into())
+            }
+        };
+
+        // If it's not long enough, then it's a runtime error that warrants an (err ..) response.
+        if name_bytes.len() < CONTRACT_MIN_NAME_LENGTH {
+            return Ok(create_principal_true_error_response(2));
+        }
+
+        // if it's too long, then this should have been caught by the type-checker
+        if name_bytes.len() > CONTRACT_MAX_NAME_LENGTH {
+            return Err(CheckErrors::TypeValueError(
+                TypeSignature::contract_name_string_ascii(),
+                name,
+            )
+            .into());
+        }
+
+        // The type-checker can't verify that the name is a valid ContractName, so we'll need to do
+        // it here at runtime.  If it's not valid, then it warrants this function evaluating to
+        // (err ..).
+        let name_string = match name {
+            // destruct again to avoid a .clone() on the inner ascii_data.data
+            Value::Sequence(SequenceData::String(CharType::ASCII(ascii_data))) => {
+                String::from_utf8(ascii_data.data).expect("FAIL: could not convert bytes of type (string-ascii 40) back to a UTF-8 string")
+            },
+            _ => {
+                unreachable!()
+            }
+        };
+
+        let contract_name = match ContractName::try_from(name_string) {
+            Ok(cn) => cn,
+            Err(_) => {
+                // not a valid contract name
+                return Ok(create_principal_true_error_response(2));
+            }
+        };
+
+        Value::Principal(PrincipalData::Contract(QualifiedContractIdentifier::new(
+            principal_data,
+            contract_name,
+        )))
+    } else {
+        // requested a standard principal
+        Value::Principal(PrincipalData::Standard(principal_data))
+    };
 
-    Ok(Value::Bool(
-        (address_is_mainnet && context_is_mainnet) || (address_is_testnet && !context_is_mainnet),
-    ))
+    if version_byte_is_valid {
+        Ok(Value::Response(ResponseData {
+            committed: version_byte_is_valid,
+            data: Box::new(principal),
+        }))
```

Comment:
```suggestion
        Value::okay(principal)
```

Reply:
`Ok(Value::okay(principal))`

Reply:
`version_byte_is_valid` may be false, so this won't do.

Reply:
Fixed